File: 114.flatten-binary-tree-to-linked-list.cpp
/*
 * @lc app=leetcode id=114 lang=cpp
 *
 * [114] Flatten Binary Tree to Linked List
 */
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start
template <typename T>
bool isLeaf(T *node)
{
    return node->left == NULL && node->right == NULL;
}

class Solution
{

private:
    pair<TreeNode *, TreeNode *> dfs(TreeNode *root)
    {
        if (root == NULL || isLeaf(root))
        {
            return {root, root};
        }

        TreeNode *curr_tail = root, *original_left = root->left, *original_right = root->right;
        root->left = NULL;
        root->right = NULL;

        if (original_left != NULL)
        {
            auto left_list = dfs(original_left);
            curr_tail->right = left_list.first;
            curr_tail = left_list.second;
        }

        if (original_right != NULL)
        {
            auto right_list = dfs(original_right);
            curr_tail->right = right_list.first;
            curr_tail = right_list.second;
        }

        return {root, curr_tail};
    }

public:
    void flatten(TreeNode *root)
    {
        dfs(root);
        return;
    }
};
// @lc code=end

File: 114.flatten-binary-tree-to-linked-list.py
#
# @lc app=leetcode id=114 lang=python3
#
# [114] Flatten Binary Tree to Linked List
#
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def isLeaf(node: TreeNode) -> bool:
    return node.left is None and node.right is None

def dfs(root: Optional[TreeNode]) -> [TreeNode, TreeNode]:
    if root is None or isLeaf(root):
        return [root, root]
    
    curr_tail = root
    original_left = root.left
    original_right = root.right
    
    if original_left is not None:
        left_list = dfs(original_left)
        curr_tail.right = left_list[0]
        curr_tail = left_list[1]
    
    if original_right is not None:
        right_list = dfs(original_right)
        curr_tail.right = right_list[0]
        curr_tail = right_list[1]
    
    return [root, curr_tail]


class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        dfs(root)
        return
        
# @lc code=end

